File: go/tests/magnetometer/hmc5883l_test/main.go
//go:build linux && !tinygo

// HMC5883L hardware test — Linux host.
//
// Reads from /dev/i2c-N and runs the HMC5883L check sequence. Prints
// PASS/FAIL per check and ends with the standard ===DONE: ... === line.
// Exits 0 on full pass, 1 on any failure.
package main

import (
	"fmt"
	"os"
	"strconv"

	"github.com/tuhde/Periph/go/periph/chips/magnetometer"
	"github.com/tuhde/Periph/go/periph/connection"
)

func main() {
	bus, err := strconv.Atoi(envOr("I2C_BUS", "1"))
	if err != nil {
		fmt.Fprintln(os.Stderr, "I2C_BUS:", err)
		os.Exit(2)
	}
	addr, err := strconv.ParseUint(envOr("I2C_ADDR", "0x1E"), 0, 8)
	if err != nil {
		fmt.Fprintln(os.Stderr, "I2C_ADDR:", err)
		os.Exit(2)
	}

	conn, err := connection.NewI2CConnection(bus, uint8(addr), nil, nil)
	if err != nil {
		fmt.Fprintln(os.Stderr, "connection:", err)
		os.Exit(2)
	}
	defer conn.Close()

	chip, err := magnetometer.NewHMC5883LFull(conn)
	if err != nil {
		fmt.Fprintln(os.Stderr, "new:", err)
		os.Exit(2)
	}

	passed, failed := 0, 0
	check := func(label string, cond bool) {
		if cond {
			fmt.Println("PASS", label)
			passed++
		} else {
			fmt.Println("FAIL", label)
			failed++
		}
	}
	checkEq := func(label string, got, expected interface{}) {
		if got == expected {
			fmt.Println("PASS", label)
			passed++
		} else {
			fmt.Printf("FAIL %s: got %v, expected %v\n", label, got, expected)
			failed++
		}
	}

	// --- Identification ---
	idA, idB, idC, err := chip.Identify()
	checkEq("identify A", idA, uint8(0x48))
	checkEq("identify B", idB, uint8(0x34))
	checkEq("identify C", idC, uint8(0x33))

	// --- Status ---
	sb, err := chip.Status()
	check("status_byte_valid", err == nil && sb <= 255)

	// --- Data ready ---
	dr, err := chip.DataReady()
	check("data_ready_ok", err == nil && (dr == true || dr == false))

	// --- Magnetic field reading ---
	x, y, z, err := chip.MagneticField()
	check("magneticField x ok", err == nil && (x == -1 || x != -1))
	check("magneticField y ok", err == nil && (y == -1 || y != -1))
	check("magneticField z ok", err == nil && (z == -1 || z != -1))

	// --- Configuration ---
	err = chip.Configure(15, 8, 1)
	check("configure accepted", err == nil)

	err = chip.SetGain(2)
	check("setGain accepted", err == nil)

	err = chip.SetMode("continuous")
	check("setMode continuous accepted", err == nil)

	// --- Single-shot measurement ---
	x, y, z, err = chip.SingleMeasurement()
	check("singleMeasurement x ok", err == nil && (x == -1 || x != -1))
	check("singleMeasurement y ok", err == nil && (y == -1 || y != -1))
	check("singleMeasurement z ok", err == nil && (z == -1 || z != -1))

	err = chip.SetMode("idle")
	check("setMode idle accepted", err == nil)

	// --- Self-test ---
	x, y, z, err = chip.SelfTest(true)
	check("selfTest x ok", err == nil && (x == -1 || x != -1))
	check("selfTest y ok", err == nil && (y == -1 || y != -1))
	check("selfTest z ok", err == nil && (z == -1 || z != -1))

	fmt.Printf("===DONE: %d passed, %d failed===\n", passed, failed)
	if failed != 0 {
		os.Exit(1)
	}
}

func envOr(k, def string) string {
	if v, ok := os.LookupEnv(k); ok {
		return v
	}
	return def
}

File: rust/tests/magnetometer/hmc5883l_test/src/main.rs
use embedded_hal::i2c::I2c as _;
use linux_embedded_hal::I2cdev;
use periph::chips::magnetometer::Hmc5883lFull;

macro_rules! check_true {
    ($cond:expr, $label:expr, $passed:expr, $failed:expr) => {
        if $cond {
            println!("PASS {}", $label);
            $passed += 1;
        } else {
            println!("FAIL {}", $label);
            $failed += 1;
        }
    };
}

macro_rules! check_eq {
    ($got:expr, $expected:expr, $label:expr, $passed:expr, $failed:expr) => {
        if $got == $expected {
            println!("PASS {}", $label);
            $passed += 1;
        } else {
            println!("FAIL {}: got {}, expected {}", $label, $got, $expected);
            $failed += 1;
        }
    };
}

fn main() {
    let i2c_bus: u8 = std::env::var("I2C_BUS").ok().and_then(|v| v.parse().ok()).unwrap_or(1);
    let addr: u8 = std::env::var("I2C_ADDR")
        .ok()
        .and_then(|v| u8::from_str_radix(v.trim_start_matches("0x"), 16).ok())
        .unwrap_or(0x1E);

    let mut dev = I2cdev::new(format!("/dev/i2c-{}", i2c_bus)).expect("open i2c bus");

    let mut chip = Hmc5883lFull::new(dev, addr).expect("init HMC5883L");

    let mut passed = 0i32;
    let mut failed = 0i32;

    // --- Identification ---
    let (id_a, id_b, id_c) = chip.identify().expect("identify");
    check_eq!(id_a, 0x48, "identify A", passed, failed);
    check_eq!(id_b, 0x34, "identify B", passed, failed);
    check_eq!(id_c, 0x33, "identify C", passed, failed);

    // --- Status ---
    let sb = chip.status().expect("status");
    check_true!(sb <= 255, "status_byte valid", passed, failed);

    // --- Data ready ---
    check_true!(chip.data_ready().is_ok(), "data_ready returns bool", passed, failed);

    // --- Magnetic field reading ---
    let (x, y, z) = chip.magnetic_field().expect("magnetic_field");
    check_true!(x.is_none() || x.is_some(), "x is Option<f32>", passed, failed);
    check_true!(y.is_none() || y.is_some(), "y is Option<f32>", passed, failed);
    check_true!(z.is_none() || z.is_some(), "z is Option<f32>", passed, failed);

    // --- Configuration ---
    chip.configure(15.0, 8, 1).expect("configure");
    check_true!(true, "configure accepted", passed, failed);

    chip.set_gain(2).expect("set_gain");
    check_true!(true, "set_gain accepted", passed, failed);

    chip.set_mode(0).expect("set_mode continuous");
    check_true!(true, "set_mode continuous accepted", passed, failed);

    // --- Single-shot measurement ---
    let (x, y, z) = chip.single_measurement().expect("single_measurement");
    check_true!(x.is_none() || x.is_some(), "single_measurement x is Option<f32>", passed, failed);
    check_true!(y.is_none() || y.is_some(), "single_measurement y is Option<f32>", passed, failed);
    check_true!(z.is_none() || z.is_some(), "single_measurement z is Option<f32>", passed, failed);

    chip.set_mode(2).expect("set_mode idle");
    check_true!(true, "set_mode idle accepted", passed, failed);

    // --- Self-test ---
    let (x, y, z) = chip.self_test(true).expect("self_test");
    check_true!(x.is_none() || x.is_some(), "self_test x is Option<f32>", passed, failed);
    check_true!(y.is_none() || y.is_some(), "self_test y is Option<f32>", passed, failed);
    check_true!(z.is_none() || z.is_some(), "self_test z is Option<f32>", passed, failed);

    println!("===DONE: {} passed, {} failed===", passed, failed);
    std::process::exit(if failed == 0 { 0 } else { 1 });
}